src/librustc/middle/resolve_lifetime.rs

```diff
@@ -375,15 +378,32 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
         match ty.node {
             hir::TyBareFn(ref c) => {
                 let scope = Scope::Binder {
-                    lifetimes: c.lifetimes.iter().map(|def| {
-                            Region::late(self.hir_map, def)
-                        }).collect(),
+                    lifetimes: c.generic_params
+                        .iter()
+                        .flat_map(|param| if let hir::GenericParam::Lifetime(ref def) = *param {
```

Comment:
Hmm, you're not using `.lifetimes()` here, but you are above, is that intentional?

Reply:
Can't because this is a `Vec<GenericParam>`, not a `Generics` instance. In my WIP rebase, this case actually came up a bunch more and I added a `GenericParmsExt` to deal with this (where I had to write out the FlatMap type bc. impl trait in traits isn't available yet), but then had to realize that it's sometimes `hir::GenericParam` and sometimes `ast::GenericParam`.

Reply:
At least in this file it's always HIR.

Reply:
I just re-checked and it's actually almost always HIR, just in the form of `P<[GenericParam]>` in some cases. Only very few things that were new when I started the rebase were in the AST (in the ast -> hir lowering specifically).

Reply:
You can impl the trait on `[GenericParam]` directly.